### PR TITLE
code cleanup: Simplify virtctl VM command files

### PR DESCRIPTION
**Pull request by prnaraya**
**What this PR does / why we need it**:

As a follow up to #9183, simplified the nested if/else logic in the Stop and Restart command. Also moved the  code block for setting the dryRunOption to a common helper function and made dryRunOption a global variable as it used in multiple different commands.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

/sig code-quality

### CODE REVIEW

File: pkg/virtctl/vm/common.go [hunk 1]
```diff
@@ -48,6 +49,7 @@ var (
 	volumeName   string
 	persist      bool
 	dryRun       bool
+	dryRunOption []string
```
Comment (EdDev) [inline pkg/virtctl/vm/common.go:52]: The usage of globals is discouraged.
It is unfortunate that we have many in-place already, but at least we better not add more.

**Review by EdDev — CHANGES_REQUESTED** [review]

File: pkg/virtctl/vm/common.go (hunk 1 above)
Comment (prnaraya) [reply pkg/virtctl/vm/common.go]: Updated to eliminate the global variable.

File: pkg/virtctl/vm/stop.go [hunk 2]
```diff
@@ -66,27 +66,29 @@ func (o *Command) stopRun(args []string, cmd *cobra.Command) error {
...
-
-	if forceRestart {
-		if cmd.Flags().Changed(gracePeriodArg) {
-			err = virtClient.VirtualMachine(namespace).ForceStop(context.Background(), vmiName, &v1.StopOptions{GracePeriod: &gracePeriod, DryRun: dryRunOption})
-			if err != nil {
-				return fmt.Errorf("Error force stopping VirtualMachine, %v", err)
-			}
-		} else if !cmd.Flags().Changed(gracePeriodArg) {
-			return fmt.Errorf("Can not force stop without gracePeriod")
+	if cmd.Flags().Changed(gracePeriodArg) {
```
Comment (EdDev) [inline pkg/virtctl/vm/stop.go:64]: I think the complexity is still there, just in a bit different order.

To reduce complexity, you could treat the errors first and only then continue.
E.g.:
```go
gpChanged = cmd.Flags().Changed(gracePeriodArg)

if gpChanged != forceRestart {
  return fmt.Errorf("Can not set gracePeriod without `--force=true`" or force stop without gracePeriod")
}

if forceRestart {
  err = virtClient.VirtualMachine(namespace).ForceStop(context.Background(), vmiName, &v1.StopOptions{GracePeriod: &gracePeriod, DryRun: dryRunOption})
 ...
} else {
  err = virtClient.VirtualMachine(namespace).Stop(context.Background(), vmiName, &v1.StopOptions{DryRun: dryRunOption})
  ...
}

fmt.Printf("VM %s was scheduled to %s\n", vmiName, o.command)
return nil
```

File: pkg/virtctl/vm/restart.go [hunk 3]
```diff
@@ -65,28 +65,29 @@ func (o *Command) restartRun(args []string, cmd *cobra.Command) error {
...
-
-	if forceRestart {
-		if cmd.Flags().Changed(gracePeriodArg) {
-			err = virtClient.VirtualMachine(namespace).ForceRestart(context.Background(), vmiName, &v1.RestartOptions{GracePeriodSeconds: &gracePeriod, DryRun: dryRunOption})
-			if err != nil {
-				return fmt.Errorf("Error restarting VirtualMachine, %v", err)
-			}
-		} else if !cmd.Flags().Changed(gracePeriodArg) {
-			return fmt.Errorf("Can not force restart without gracePeriod")
+	if cmd.Flags().Changed(gracePeriodArg) {
```
Comment (EdDev) [inline pkg/virtctl/vm/restart.go:63]: Same here, the complexity can be reduced by treating the input validation errors first, then do the actions.

**Review by EdDev — CHANGES_REQUESTED** [review]
Could you please check and confirm that the changed areas, especially in the first two commits are covered by unit tests? If they are not, please add tests to cover these cases so we know nothing unexpected happened.

If we are talking about cleanup, regarding the commit messages, please make sure the subject line is not too long.
See the ["The seven rules of a great Git commit message"](https://cbea.ms/git-commit/). It is better not to pass the 72 chars so the line will not break.

File: pkg/virtctl/vm/stop.go (hunk 2 above)
Comment (prnaraya) [reply pkg/virtctl/vm/stop.go]: Done

File: pkg/virtctl/vm/restart.go (hunk 3 above)
Comment (prnaraya) [reply pkg/virtctl/vm/restart.go]: Done

**Comment by prnaraya** [comment]
@EdDev I checked and there were indeed cases in that block of code that weren't being covered by unit tests. I've added some in.

File: pkg/virtctl/vm/stop.go [hunk 4]
```diff
@@ -65,19 +64,20 @@ func (o *Command) stopRun(args []string, cmd *cobra.Command) error {
...
+	
+	gracePeriodChanged := cmd.Flags().Changed(gracePeriodArg)
 
-	if cmd.Flags().Changed(gracePeriodArg) && forceRestart == false {
-		return fmt.Errorf("Can not set gracePeriod without --force=true")
+	if gracePeriodChanged != forceRestart {
+		if !forceRestart {
+			return fmt.Errorf("Can not set gracePeriod without --force=true")
+		}
+		return fmt.Errorf("Can not force stop without gracePeriod")
```
Comment (EdDev) [inline pkg/virtctl/vm/stop.go:69]: I do not think we need to be that verbose about the error, it could be just one combined message.
But this is a nit.

File: pkg/virtctl/vm/stop.go [hunk 5]
```diff
@@ -87,6 +87,5 @@ func (o *Command) stopRun(args []string, cmd *cobra.Command) error {
 	}
 
 	fmt.Printf("VM %s was scheduled to %s\n", vmiName, o.command)
-
```
Comment (EdDev) [inline pkg/virtctl/vm/stop.go:90]: nit: unrelated change

File: pkg/virtctl/vm/vm_test.go [hunk 6]
```diff
@@ -273,35 +273,81 @@ var _ = Describe("VirtualMachine", func() {
...
 			vm := kubecli.NewMinimalVM(vmName)
 
-			kubecli.MockKubevirtClientInstance.EXPECT().VirtualMachine(k8smetav1.NamespaceDefault).Return(vmInterface).Times(1)
-			gracePeriod := int64(0)
-			restartOptions := v1.RestartOptions{
-				GracePeriodSeconds: &gracePeriod,
-				DryRun:             nil,
-			}
-			vmInterface.EXPECT().ForceRestart(context.Background(), vm.Name, &restartOptions).Return(nil).Times(1)
+			if gracePeriodSet != force {
```
Comment (EdDev) [inline pkg/virtctl/vm/vm_test.go:280]: There is an attempt here to use a common function when it is clearly not a fit.
You need two functions here, one that treats `gracePeriodSet != force` scenarios and one for the other.

A test table is a fit when multiple scenarios/inputs share the same test logic, but here it is clearly not the case.

File: pkg/virtctl/vm/vm_test.go [hunk 7]
```diff
@@ -273,35 +273,81 @@ var _ = Describe("VirtualMachine", func() {
...
-			cmd := clientcmd.NewVirtctlCommand("stop", vmName, "--force", "--grace-period=0")
+			cmd := clientcmd.NewVirtctlCommand("stop", vmName)
 			Expect(cmd.Execute()).To(Succeed())
 		})
+
+		DescribeTable("with flags:", func(force bool, gracePeriodSet bool) {
+			var cmd *cobra.Command
+			vm := kubecli.NewMinimalVM(vmName)
+
+			if gracePeriodSet != force {
```
Comment (EdDev) [inline pkg/virtctl/vm/vm_test.go:324]: Same here.

**Review by EdDev — CHANGES_REQUESTED** [review]
I had some nits about the production code, none of them is a blocker for me.

However, the unit tests do have a problem IMO, see the inline comment about them.

Thank you for investing in this.

File: pkg/virtctl/vm/vm_test.go (hunk 6 above)
Comment (prnaraya) [reply pkg/virtctl/vm/vm_test.go]: Updated and split up the tests.

File: pkg/virtctl/vm/stop.go (hunk 4 above)
Comment (prnaraya) [reply pkg/virtctl/vm/stop.go]: I consolidated this into a single error message.

**Review by EdDev — APPROVED** [review]
Thank you!

File: pkg/virtctl/vm/restart.go [hunk 8]
```diff
@@ -64,26 +64,23 @@ func (o *Command) restartRun(args []string, cmd *cobra.Command) error {
...
-			return fmt.Errorf("Can not force restart without gracePeriod")
+		err = virtClient.VirtualMachine(namespace).ForceRestart(context.Background(), vmiName, &v1.RestartOptions{GracePeriodSeconds: &gracePeriod, DryRun: dryRunOption})
+		if err != nil {
+			return fmt.Errorf("Error force stopping VirtualMachine, %v", err)
 		}
 	} else {
 		err = virtClient.VirtualMachine(namespace).Restart(context.Background(), vmiName, &v1.RestartOptions{DryRun: dryRunOption})
 		if err != nil {
-			return fmt.Errorf("Error restarting VirtualMachine %v", err)
+			return fmt.Errorf("Error stopping VirtualMachine %v", err)
```
Comment (acardace) [inline pkg/virtctl/vm/restart.go:76]: why stopping? shouldn't it be `restarting`?

Comment (prnaraya) [reply pkg/virtctl/vm/restart.go]: That's correct, good catch. I've fixed it.

**Comment by EdDev** [comment]
> @EdDev @acardace I've made a small update to this PR, would you be able to take a look? Thanks in advance.

@prnaraya , it looks like you changes something but also rebased in the same push.
This makes it very hard to identify what you changed and requires a full re-review.
If you have a diff link showing what you changed from the last lgtm, please share it.

It will help the reviewers if you do not rebase and change in a single push, but do these in separate steps (e.g. change push, rebase push).

**Comment by EdDev** [comment]
Seems that this is the change: https://github.com/kubevirt/kubevirt/compare/ed2546cd6b975b734a8c9a0863c5a952dd7b2d4a..e9d1a3c12b9c447c4889ecaa8101f2e7c9b1e832#diff-e284c5fce546e3ad989c61af6278fbd8a619993d966d8d1925c3ac0fac12f37c

I hope I did not miss anything else.

/lgtm

**Review by jean-edouard — COMMENTED** [review]
/approve
Just one note, it looks like `--force` is redundant with `--grace-period`, but I think you're well aware and already looking into this! Thanks
